packages/connection-storage/src/connection-storage.ts

```diff
@@ -98,11 +107,128 @@ export class ConnectionStorage {
     this.calledOnce = true;
   }
 
-  private static mapStoredConnectionToConnectionInfo(
-    storedConnectionInfo: ConnectionInfo,
+  static async migrateToSafeStorage() {
+    // If user lands on this version of Compass with legacy connections (using storage mixin),
+    // we will ignore those and only migrate connections that have connectionInfo.
+    const connections = (await this.getConnections()).filter(
+      (x) => x.connectionInfo
+    );
+    if (connections.length === 0) {
+      log.info(
+        mongoLogId(1_001_000_200),
+        'Connection Storage',
+        'No connections to migrate'
+      );
+      return;
+    }
+
+    if (
+      connections.filter((x) => 'version' in x && x.version === this.version)
+        .length === connections.length
+    ) {
+      log.info(
+        mongoLogId(1_001_000_205),
+        'Connection Storage',
+        'Connections already migrated'
+      );
+      return;
+    }
+
+    // Get the secrets from keychain. If there're no secrets, we still want to
+    // migrate the connects to a new version.
+    const secrets = await this.getKeytarCredentials();
+
+    // Connections that are not migrated yet or that failed to migrate,
+    // are not expected to have a version property.
+    const unmigratedConnections = connections.filter((x) => !('version' in x));
+    log.info(
+      mongoLogId(1_001_000_206),
+      'Connection Storage',
+      'Migrating connections',
+      {
+        num_connections: unmigratedConnections.length,
+        num_secrets: Object.keys(secrets).length,
+      }
+    );
+
+    let numFailedToMigrate = 0;
+
+    for (const { connectionInfo } of unmigratedConnections) {
+      // We have connectionInfo as we are limiting this migration to
+      // connections that have connectionInfo (not legacy connections)
+      const _id = connectionInfo!.id;
```

Comment:
Would we still want to provide a more explicit error message in this case than something like `cannot read property id of undefined`?

Reply:
the connections list is already being [filtered](https://github.com/mongodb-js/compass/pull/5115/files/7cd05a37bf6c506bd960844c80f514573745ed85#diff-11d5f2650641bd452506da099a11965f742934b623ee0aebcd6028658934a419R113) as we fetch it ensuring that we only consider connections that have `connectionInfo` property.

Reply:
Yeah, good point – I’ve left a comment above that allows you to get rid of the non-null assertion here by fully typing this out, if you like, otherwise feel free to just resolve